File: quartic_solver.py
import numpy as np

def _get_quartic_solution(W, pm1=+1, pm2=+1):
    def isclose(a, b, rel_tol=1e-09, abs_tol=1e-8):
        return np.abs(a - b) <= np.maximum(
            rel_tol * np.maximum(np.abs(a), np.abs(b)), np.array(abs_tol)
        )

    def isnonzero(a):
        return not isclose(a, np.array(0)).item()

    W_conj = np.conj(W)

    A = (1 - W * W_conj) ** 3 / 27 + (W**2 - W_conj**2) ** 2 / 16
    C = (2 * (W_conj**2 - W**2) + 8 * np.sqrt(A)) ** (1 / 3)
    if isnonzero(C):
        u1 = C - 4 * (1 - W * W_conj) / (3 * C)
    else:
        u1 = (4 * (W_conj**2 - W**2)) ** (1 / 3)

    term1 = W + pm1 * np.sqrt(u1 + W**2)
    term2 = u1 + pm1 * (
        ((2 * W_conj + W * u1) / (np.sqrt(u1 + W**2)))
        if isnonzero(u1 + W**2)
        else np.sqrt(u1**2 + 4)
    )

    z = (term1 + pm2 * np.sqrt(term1**2 - 2 * term2)) / 2

    # if magnitude of z is not close to 1 return None
    if not isclose(np.abs(z), np.array(1), rel_tol=1e-09, abs_tol=1e-8):
        return None

    return z


def get_ACLE_angular_solns(W):
    solns = set()
    
    for i in range(4):
        # TODO can be sped by restructuring loop
        pm1 = [+1, -1][i // 2]
        pm2 = [+1, -1][i % 2]
        s = _get_quartic_solution(W, pm1, pm2)
        if s is not None:
            solns.add(s)
    return solns


if __name__ == "__main__":
    W = complex(0.03, 0.1)
    solutions = get_ACLE_angular_solns(W)

    for i, sol in enumerate(solutions):
        print(f"Solution {i+1}: {sol:.3f}")

File: quartic_solver.cpp
#include <iostream>
#include <complex>
#include <cmath>
#include <vector>
#include <iomanip> // Include the necessary header for std::setprecision
#include "quartic_solver.h" // Include the header file for the function _get_quartic_solution

using namespace std;

std::complex<double> _get_quartic_solution(std::complex<double> W, int pm1 = +1, int pm2 = +1) {
    auto isclose = [](std::complex<double> a, std::complex<double> b, double rel_tol = 1e-09, double abs_tol = 1e-8) {
        return std::abs(a - b) <= std::max(rel_tol * std::max(std::abs(a), std::abs(b)), abs_tol);
    };

    std::complex<double> x = 0;

    std::complex<double> W_conj = std::conj(W);

    std::complex<double> A = std::pow((1.0 - W * W_conj), 3) / 27.0 + std::pow((W * W - W_conj * W_conj), 2) / 16.0;
    std::complex<double> C = std::pow((2.0 * (W_conj * W_conj - W * W) + 8.0 * std::sqrt(A)), 1.0 / 3.0);

    std::complex<double> u1;
    if (!isclose(C, std::complex<double>(0.0, 0.0))) {
        u1 = C - 4.0 * (1.0 - W * W_conj) / (3.0 * C);
    } else {
        u1 = std::pow((4.0 * (W_conj * W_conj - W * W)), 1.0 / 3.0);
    }

    std::complex<double> term1 = W + static_cast<double>(pm1) * std::sqrt(u1 + W * W);
    std::complex<double> term2 = u1 + static_cast<double>(pm1) * (((2.0 * W_conj + W * u1) / std::sqrt(u1 + W * W)) != 0.0 ? ((2.0 * W_conj + W * u1) / std::sqrt(u1 + W * W)) : std::sqrt(u1 * u1 + 4.0));

    std::complex<double> z = (term1 + static_cast<double>(pm2) * std::sqrt(term1 * term1 - 2.0 * term2)) / 2.0;

    // if magnitude of z is not close to 1 return None
    if (!isclose(std::abs(z), 1.0, 1e-09, 1e-8)) {
        return std::complex<double>(0.0, 0.0);
    }

    return z;
}

std::vector<std::complex<double>> get_ACLE_angular_solns(std::complex<double> W) {
    std::vector<std::complex<double>> solns;

    for (int i = 0; i < 4; i++) {
        int pm1 = (i / 2 == 0) ? +1 : -1;
        int pm2 = (i % 2 == 0) ? +1 : -1;
        std::complex<double> s = _get_quartic_solution(W, pm1, pm2);
        if (s != std::complex<double>(0.0, 0.0)) {
            solns.push_back(s);
        }
    }
    return solns;
}


int main() {
    std::complex<double> W(0.03, 0.1);
    std::vector<std::complex<double>> solutions = get_ACLE_angular_solns(W);

    for (int i = 0; i < solutions.size(); i++) {
        std::cout << "Solution " << i + 1 << ": " << std::fixed << std::setprecision(3) << solutions[i] << std::endl;
    }

    return 0;
}
